identity-account-storage/src/stronghold/error.rs

```diff
@@ -1,37 +1,104 @@
 // Copyright 2020-2022 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-pub type IotaStrongholdResult<T> = Result<T, StrongholdError>;
+use std::fmt::Display;
+
+use iota_stronghold::procedures::ProcedureError;
+use iota_stronghold::ClientError;
+use iota_stronghold::MemoryError;
+use iota_stronghold::SnapshotPath;
+
+use crate::types::KeyLocation;
+
+use super::client_path::ClientPath;
+
+pub(crate) type StrongholdResult<T> = Result<T, StrongholdError>;
+pub(crate) type ProcedureName = &'static str;
 
 /// Caused by errors from the [`iota_stronghold`] crate.
+#[non_exhaustive]
 #[derive(Debug, thiserror::Error, strum::IntoStaticStr)]
 pub enum StrongholdError {
-  #[error(transparent)]
-  StrongholdActorError(#[from] iota_stronghold::ActorError),
-  #[error(transparent)]
-  StrongholdWriteError(#[from] iota_stronghold::WriteError),
-  #[error(transparent)]
-  StrongholdReadError(#[from] iota_stronghold::ReadError),
-  #[error(transparent)]
-  StrongholdFatalEngineError(#[from] iota_stronghold::FatalEngineError),
-  #[error(transparent)]
-  StrongholdMailboxError(#[from] iota_stronghold::MailboxError),
-  /// Caused by a failing Stronghold procedure.
-  #[error("Stronghold procedure failed: {0}")]
-  StrongholdProcedureFailure(#[from] iota_stronghold::procedures::ProcedureError),
-  /// Caused by attempting to perform an invalid IO operation.
-  #[error(transparent)]
-  IoError(#[from] std::io::Error),
-
-  /// Caused by attempting to access a Stronghold snapshot without a password.
-  #[error("Stronghold snapshot password not found")]
-  StrongholdPasswordNotSet,
-  /// Caused by errors from an invalid Stronghold procedure.
-  #[error("Stronghold error: {0}")]
-  StrongholdResult(String),
-  #[error("Record Error")]
-  RecordError,
-  /// Caused by attempting to parse an invalid Stronghold resource index.
-  #[error("Stronghold resource index malformed")]
-  InvalidResourceIndex,
+  #[error("failed to `{0}` stronghold client `{1}`")]
+  Client(ClientOperation, ClientPath, #[source] ClientError),
+  #[error("store `{0}` operation failed")]
+  Store(StoreOperation, #[source] ClientError),
+  #[error("vault operation `{0}` failed")]
+  Vault(VaultOperation, #[source] ClientError),
+  #[error("procedure `{0}` operating on locations {1:?} failed")]
+  Procedure(ProcedureName, Vec<KeyLocation>, #[source] ProcedureError),
```

Comment:
Not all procedures will be dealing with `KeyLocation`, since this is either an `Ed25519` or an `X25519` Key for now. For example, when dealing with the procedures related to shared secrets the type is neither. However, we are dealing with random locations in that case. Should we change `KeyLocation` or is it fine to have an empty location in these cases?

Reply:
Would changing it to `Vec<String>` account for shared secrets? I think `Location` alone doesn't have a string representation.

Reply:
If the need arises to return errors that include non-`KeyLocation`s, then Craig's suggestion is great. Though, I think for debugging it should be helpful enough to include the `KeyLocation`s, as the other, temporary locations will probably be completely random or static, and we'll be able to trace the issue based on the procedure name alone, I imagine. From how I understand `execute_procedure_chained`, it will delete those temp locations by itself, so we shouldn't have to deal with potentially failing `RevokeData` procedures, either: https://github.com/iotaledger/stronghold.rs/blob/dev-refactor/client/src/types/client.rs#L247-#L254

Reply:
> temporary locations will probably be completely random or static, and we'll be able to trace the issue based on the procedure name alone, I imagine.

I thought so too. For now, I won't change the type, but will change it to `String` in case I need it.